File: 54-SpiralMatrix/54-SpiralMatrix.py
# Last updated: 9/16/2025, 12:53:25 AM
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        sr = 0
        er = len(matrix) -1
        sc = 0
        ec = len(matrix[0]) -1
        a = []
        while(sr<=er and sc<=ec):

            for j in range (sc, ec+1):
                a.append(matrix[sr][j])
            sr +=1

            if sr<=er:
                for i in range(sr, er+1):
                    a.append(matrix[i][ec])
                ec -=1

            if sc<=ec and sr<= er:
                for j in range(ec, sc-1, -1):
                    a.append(matrix[er][j])
                er -=1

            if sr <= er and sc <= ec:
                for i in range(er, sr-1, -1):
                    a.append(matrix[i][sc])
                sc+=1
            
        return a
            

File: 54-SpiralMatrix/54-SpiralMatrix.cpp
// Last updated: 2/13/2026, 12:58:17 PM
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) 
    {
        int n = matrix.size();
        int m = matrix[0].size();
        int t = 0, b = n-1, l = 0, r = m-1;

        vector<int> ans;

        while(l<=r && t<=b)
        {
            for(int i = l; i<=r; i++)
                ans.push_back(matrix[t][i]);
            t++;

            for(int j = t; j<=b; j++)
                ans.push_back(matrix[j][r]);
            r--;

            if (t<=b)
            {
                for(int i = r; i>=l; i--)
                    ans.push_back(matrix[b][i]);
                b--;
            }

            if(l<=r)
            { 
                for(int j = b; j>=t; j--)
                ans.push_back(matrix[j][l]);
                l++;  
            }
            
        }
        return ans;
    }
};